database/database.go

```diff
@@ -486,3 +512,74 @@ func (s *DatabaseService) DeleteExecutionPayloads(idFirst, idLast uint64) error
+	entries := []*BlockBuilderEntry{}
+	err := s.DB.Select(&entries, query, collateralID)
+	out := []string{}
+	for _, v := range entries {
+		out = append(out, v.BuilderPubkey)
+	}
+	return out, err
+}
+
+func (s *DatabaseService) UpsertBuilderDemotion(submitBlockRequest *types.BuilderSubmitBlockRequest, signedBeaconBlock *types.SignedBeaconBlock, signedValidatorRegistration *types.SignedValidatorRegistration) error {
+	if submitBlockRequest == nil {
+		return fmt.Errorf("nil submitBlockRequest invalid for UpsertBuilderDemotion")
+	}
+
+	_submitBlockRequest, err := json.Marshal(submitBlockRequest)
+	if err != nil {
+		return err
+	}
+	bidTrace := submitBlockRequest.Message
+	builderDemotionEntry := BuilderDemotionEntry{
+		SubmitBlockRequest: NewNullString(string(_submitBlockRequest)),
+
+		Slot:  bidTrace.Slot,
+		Epoch: bidTrace.Slot / uint64(common.SlotsPerEpoch),
+
+		BuilderPubkey:  bidTrace.BuilderPubkey.String(),
+		ProposerPubkey: bidTrace.ProposerPubkey.String(),
+
+		Value: bidTrace.Value.String(),
+
+		BlockHash: bidTrace.BlockHash.String(),
+	}
+
+	if signedBeaconBlock != nil {
+		_signedBeaconBlock, err := json.Marshal(signedBeaconBlock)
+		if err != nil {
+			return err
+		}
+		_signedValidatorRegistration, err := json.Marshal(signedValidatorRegistration)
+		if err != nil {
+			return err
+		}
+		builderDemotionEntry.SignedBeaconBlock = NewNullString(string(_signedBeaconBlock))
+		builderDemotionEntry.SignedValidatorRegistration = NewNullString(string(_signedValidatorRegistration))
+	}
+
+	var query string
+	// If block_hash + builder_pubkey conflicts and we have a published block, fill in fields needed for the refund.
+	if signedBeaconBlock != nil {
+		query = `INSERT INTO ` + vars.TableBuilderDemotions + `
+			(submit_block_request, signed_beacon_block, signed_validator_registration, slot, epoch, builder_pubkey, proposer_pubkey, value, fee_recipient, gas_limit, block_hash) VALUES
+			(:submit_block_request, :signed_beacon_block, :signed_validator_registration, :slot, :epoch, :builder_pubkey, :proposer_pubkey, :value, :fee_recipient, :gas_limit, :block_hash)
+			ON CONFLICT (block_hash, builder_pubkey) DO UPDATE SET
+				signed_beacon_block = :signed_beacon_block,
+				signed_validator_registration = :signed_validator_registration,
+				fee_recipient = :fee_recipient,
+				gas_limit = :gas_limit;
+			`
+	} else {
+		// If the block_hash + builder_pubkey conflicts, then all the relevant data must be there already, so do nothing.
+		query = `INSERT INTO ` + vars.TableBuilderDemotions + `
+			(submit_block_request, signed_beacon_block, signed_validator_registration, slot, epoch, builder_pubkey, proposer_pubkey, value, fee_recipient, gas_limit, block_hash) VALUES
+			(:submit_block_request, :signed_beacon_block, :signed_validator_registration, :slot, :epoch, :builder_pubkey, :proposer_pubkey, :value, :fee_recipient, :gas_limit, :block_hash)
+			ON CONFLICT (block_hash, builder_pubkey) DO NOTHING`
```

Comment:
Consider abstracting these 4 lines (without `NOTHING`) into a variable which is used twice (for the two `query`).